sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedReadFromBoundedSource.java

```diff
@@ -0,0 +1,341 @@
+        return new CheckpointCoder<>(components.get(0));
+      }
+
+      private final Coder<List<TimestampedValue<T>>> elemsCoder;
+      private final Coder<T> elemCoder;
+      @SuppressWarnings("rawtypes")
+      private final SerializableCoder<BoundedSource> serializableCoder =
+          SerializableCoder.of(BoundedSource.class);
+
+      CheckpointCoder(Coder<T> elemCoder) {
+        this.elemsCoder = ListCoder.of(TimestampedValue.TimestampedValueCoder.of(elemCoder));
+        this.elemCoder = elemCoder;
+      }
+
+      @Override
+      public void encode(Checkpoint<T> value, OutputStream outStream, Context context)
+          throws CoderException, IOException {
+        Context nested = context.nested();
+        elemsCoder.encode(value.residualElements, outStream, nested);
+        serializableCoder.encode(value.residualSource, outStream, nested);
+      }
+
+      @SuppressWarnings("unchecked")
+      @Override
+      public Checkpoint<T> decode(InputStream inStream, Context context)
+          throws CoderException, IOException {
+        Context nested = context.nested();
+        return new Checkpoint<>(
+            elemsCoder.decode(inStream, nested),
+            serializableCoder.decode(inStream, nested));
+      }
+
+      @Override
+      public List<Coder<?>> getCoderArguments() {
+        return Arrays.<Coder<?>>asList(elemCoder);
+      }
+
+      @Override
+      public void verifyDeterministic() throws NonDeterministicException {
+        serializableCoder.verifyDeterministic();
+        elemsCoder.verifyDeterministic();
+      }
+    }
+
+    @VisibleForTesting
+    class Reader extends UnboundedReader<T> {
+      private final PipelineOptions options;
+
+      private final List<TimestampedValue<T>> residualElements;
+      private @Nullable BoundedSource<T> residualSource;
+
+      private boolean currentElementInList;
+      private int currentIndexInList;
+      private @Nullable BoundedReader<T> reader;
+      private boolean done;
+
+      public Reader(
+          List<TimestampedValue<T>> residualElements,
+          @Nullable BoundedSource<T> residualSource,
+          PipelineOptions options) {
+        this.residualElements = Preconditions.checkNotNull(residualElements, "residualElements");
+        this.residualSource = residualSource;
+        this.options = Preconditions.checkNotNull(options, "options");
+        this.currentElementInList = true;
+        this.currentIndexInList = -1;
+        this.reader = null;
+        this.done = false;
+      }
+
+      @Override
+      public boolean start() throws IOException {
+        return advance();
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        if (advanceInList()) {
+          done = false;
+        } else {
+          done = !advanceInReader();
+        }
+        if (done) {
+          close();
+        }
+        return !done;
+      }
+
+      private boolean advanceInList() {
```

Comment:
can you comment on what the semantics of the Reader are? What state it tracks, how it operates, how it checkpoints, etc.? Need a detailed explanation of how UnboundedReadFromBoundedSource works and why it is correct.